server/src/main/java/org/elasticsearch/search/aggregations/support/BuiltinValuesSourceType.java

```diff
@@ -0,0 +1,214 @@
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            // TODO: Implement this or get rid of ANY
+            throw new UnsupportedOperationException("ValuesSourceType.ANY is still a special case");
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            return BYTES.replaceMissing(valuesSource, rawMissing, docValueFormat, now);
+        }
+    },
+    NUMERIC {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.Numeric.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return new ValuesSource.Numeric.Script(script, scriptValueType);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+
+            if (!(fieldContext.indexFieldData() instanceof IndexNumericFieldData)) {
+                // TODO: Is this the correct exception type here?
+                throw new IllegalArgumentException("Expected numeric type on field [" + fieldContext.field() +
+                    "], but got [" + fieldContext.fieldType().typeName() + "]");
+            }
+
+            ValuesSource.Numeric dataSource = new ValuesSource.Numeric.FieldData((IndexNumericFieldData)fieldContext.indexFieldData());
+            if (script != null) {
+                // What's the difference between ValuesSource.Numeric.Script and ValuesSource.Numeric.WithScript?
+                dataSource = new ValuesSource.Numeric.WithScript(dataSource, script);
+            }
+            return dataSource;
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            Number missing = docValueFormat.parseDouble(rawMissing.toString(), false, now);
+            return MissingValues.replaceMissing((ValuesSource.Numeric) valuesSource, missing);
+        }
+    },
+    BYTES {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.Bytes.WithOrdinals.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return new ValuesSource.Bytes.Script(script);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            final IndexFieldData<?> indexFieldData = fieldContext.indexFieldData();
+            ValuesSource dataSource;
+            if (indexFieldData instanceof IndexOrdinalsFieldData) {
```

Comment:
Not important right now, but I wonder if we should try to encode some of these attributes with properties instead, in a future iteration?  `hasOrdinals()`, `isNumeric()`, `isRange()` (or `hasHistogram()` in reference to Ignacio's PR) etc, that way we can avoid doing instanceof checks?

Might not be worth it, just a thought since instanceof's always sorta make me uneasy :) And it might make things more pluggable in the future since we won't have hardcoded classes in Core, just attributes

Reply:
I'm +1 on this, but it seems out of scope for this PR.  I can open an issue for it.